[WebGPU] Flaky failures on CTS due to reaching command buffer limit

mwyrzykowski opened the pull request:
#### ab50f80f0e54c56a17c4b273bcec57dc6508085c
<pre>
[WebGPU] Flaky failures on CTS due to reaching command buffer limit
<a href="https://bugs.webkit.org/show_bug.cgi?id=275554">https://bugs.webkit.org/show_bug.cgi?id=275554</a>
&lt;radar://129971928&gt;

Reviewed by Tadeu Zagallo.

<a href="https://github.com/gpuweb/gpuweb/issues/4622">https://github.com/gpuweb/gpuweb/issues/4622</a> is resolved but the CTS
still needs to be updated and the CTS can create several thousand
open GPUCommandEncoder instances without calling GPUQueue.submit() on
the buffer.

Raise the limit for the CTS so we don&apos;t observe flaky failures.

* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::maxCommandBufferCount):
(WebGPU::Device::create):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::commandBufferWithDescriptor):

Canonical link: <a href="https://commits.webkit.org/280072@main">https://commits.webkit.org/280072@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/11d9b3fcfa4782f7aa5f7983c88d1f5449502dde

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55666 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34989 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8133 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58650 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6096 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42611 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6295 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44832 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4193 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57695 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32889 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47979 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25960 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29676 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4240 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51650 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5575 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60241 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5709 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52258 "Passed tests") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48049 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51746 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12332 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30820 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31905 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32986 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31652 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -124,7 +124,11 @@
     if (!isValid())
         return std::make_pair(nil, nil);
 
+#if PLATFORM(IOS)
```
Why don't we want to raise the limit on iOS?

mwyrzykowski replied:
iOS will jetsam around 2000 buffers with no other resource usage. So we will need to do something else, though possibly the CTS gets updated prior to that

tadeuzagallo replied:
I'm a bit confused by this too... are we doing this to pass the CTS test, which is incorrect/outdated?

mwyrzykowski replied:
Yes that is correct. Application should not create more than 1000 per https://github.com/gpuweb/gpuweb/issues/4622 or risk device lost. CTS is not handling device lost (it should see the device is lost, pause the test, re-create the device, resume the test cases).

mwyrzykowski replied:
What happens today is too many test cases run at once, CTS never calls GPUCommandEncoder.finish or GPUQueue.submit on the invalid encoders, we retain the MTLCommandBuffer reference until the JS object gets GC'ed from the test.

At 1000 we would currently return nil and report device lost, so those test cases failed. 

In a regular application, we see the count around 2-3 or maybe 5.

tadeuzagallo replied:
I see, and I assume `Device::maxCommandBufferCount()` is bigger than 1000, which is causing the flakiness? Should `Device::maxCommandBufferCount()` be updated to be 1000?

mwyrzykowski replied:
The flakiness occurs because the CTS does not handle device lost. If we set maxCommandBufferCount to 1000 we would still see the issue, as we would trigger a device lost but the CTS does not respond to that.

tadeuzagallo replied:
Oh! I get it now, that makes sense, thanks for walking me through it 😅

mwyrzykowski replied:
No problem :)

mwyrzykowski replied:
I actually should have investigated this more, we reached the MTLCommandBuffer limit because invalid CommandEncoder instances were retaining their MTLCommandBuffer due to incorrect logic in CommandEncoder::finish.

Increasing this limit is therefore not needed, this change can be reverted

mwyrzykowski committed:
[WebGPU] Flaky failures on CTS due to reaching command buffer limit
https://bugs.webkit.org/show_bug.cgi?id=275554
<radar://129971928>

Reviewed by Tadeu Zagallo.

https://github.com/gpuweb/gpuweb/issues/4622 is resolved but the CTS
still needs to be updated and the CTS can create several thousand
open GPUCommandEncoder instances without calling GPUQueue.submit() on
the buffer.

Raise the limit for the CTS so we don't observe flaky failures.

* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::maxCommandBufferCount):
(WebGPU::Device::create):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::commandBufferWithDescriptor):

Canonical link: https://commits.webkit.org/280072@main
